README
  -------------------------------------------------
    Linear Feedback Shift Registers for GNU Guile
  -------------------------------------------------

This  library  contains  three  scheme  modules targetted  at  GNU  Guile.  The
(communication  lfsr)  module implements  linear  feedback  shift registers  as
infinite  streams of  bits.  LFSRs  are used,  for  example,  to generate  chip
sequences  in CDMA  systems, for  synchronising preambles  or as  pseudo random
number generators.  The module  implements LFSRs in  both Fibonacci  and Galois
configurations.

The (communication  gps-ca) module,  which uses  the former  implements streams
that  produce pseudo  random gold  chip sequences  as used  in C/A  of the  GPS
satellite system. It uses LFSRs in Fibonacci configuration.

The (communication  ieee802-15-4a-preamble) module implements  ternary preamble
sequences as used in  ultra-wideband transmissions the ieee-802.15.4a standard.
The generators for these sequences use LFSRs in Galois configuration.

To run  the test-suite of the  library, the following additional  dependency is
required:   https://gitlab.com/ft/guile-tap


CHANGES
------------------------------------
 User-visible changes in guile-lfsr
------------------------------------

* v0.1.1 → v0.1.2 (released 2021-07-05):

** Bugfixes

  - Install copy of GPL to complete LGPL document.

** Diffstat

    tools/install | 3 ++-
    1 file changed, 2 insertions(+), 1 deletion(-)


* v0.1.0 → v0.1.1 (released 2021-07-04):

** Miscellaneous

  - Install package documentation to DOCDIR

** Diffstat

    Makefile      |  5 +++--
    tools/install | 34 +++++++++++++++++++++++++++++-----
    2 files changed, 32 insertions(+), 7 deletions(-)


* scratch → v0.1.0 (released 2021-07-04):

** Diffstat

    .gitignore                                    |  13 +
    LICENCE                                       | 165 +++++
    LICENCE.GPL                                   | 674 ++++++++++++++++++
    Makefile                                      |  50 ++
    README                                        |  21 +
    scheme/communication/gps-ca.scm               |  90 +++
    .../communication/ieee802-15-4a-preamble.scm  | 125 ++++
    scheme/communication/lfsr.scm                 |  84 +++
    scheme/test/setup.scm                         |  15 +
    tests/gps-ca-scm.t                            |  34 +
    tests/ieee802-15-4a-preamble-scm.t            |  58 ++
    tests/lfsr-scm.t                              |  25 +
    tests/scheme.env                              |   7 +
    tools/dev-env                                 |  11 +
    tools/guile-in-here                           |   5 +
    tools/install                                 | 171 +++++
    tools/run-single-test                         |  12 +
    17 files changed, 1560 insertions(+)


scheme/communication/lfsr.scm
;; Copyright © 2017 Frank Terbeck <ft@bewatermyfriend.org>
;;
;; This library  is free  software; you  can redistribute  it and/or  modify it
;; under the terms of the GNU Lesser General Public License as published by the
;; Free  Software Foundation;  either version  3 of  the License,  or (at  your
;; option) any later version.
;;
;; This library is distributed in the hope  that it will be useful, but WITHOUT
;; ANY  WARRANTY;  without even  the  implied  warranty of  MERCHANTABILITY  or
;; FITNESS FOR A PARTICULAR PURPOSE. See  the GNU Lesser General Public License
;; for more details.
;;
;; You should  have received a  copy of the  GNU Lesser General  Public License
;; along with  this library;  if not,  write to  the Free  Software Foundation,
;; Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
;;
;; Commentary:
;;
;; This module implements  linear feedback shift registers  as infinite streams
;; of bits.  LFSRs are used,  for example, to  generate chip sequences  in CDMA
;; systems, for synchronising preambles or as pseudo random number generators.

(define-module (communication lfsr)
  #:use-module (srfi srfi-41)
  #:use-module ((srfi srfi-151) #:prefix srfi151:)
  #:export (make-lfsr-stream-fibonacci
            make-lfsr-stream-galois
            word->bit-lfsr
            multi-xor
            reverse-initial-state
            reverse-generator-polynom
            reverse-gp-and-state))

(define (reverse-generator-polynom gp)
  (srfi151:list->bits (srfi151:bits->list gp)))

(define (reverse-initial-state state len)
  (srfi151:list->bits (srfi151:bits->list state len)))

(define (reverse-gp-and-state gp state)
  (let ((lfsr-length (- (length (srfi151:bits->list gp)) 1)))
    (cons (reverse-generator-polynom gp)
          (reverse-initial-state state lfsr-length))))

(define (multi-xor x)
  (modulo (srfi151:bit-count x) 2))

(define lfsr-iterator
  (stream-lambda (fnc head) (stream-cons head (lfsr-iterator fnc (fnc head)))))

(define (minimum-word-width word)
  (let loop ((x word) (steps 0))
    (if (zero? x)
        (- steps 1)
        (loop (ash x -1) (+ 1 steps)))))

(define (lfsr-step-fibonacci gp)
  (let* ((width (minimum-word-width gp))
         (msb (ash 1 (- width 1))))
    (lambda (x)
      (let ((new (multi-xor (logand x gp)))
            (shifted (ash x -1)))
        (if (zero? new)
            shifted
            (logior shifted msb))))))

(define (lfsr-step-galois gp)
  (lambda (x)
    (let* ((new (logand x 1)))
      (if (zero? new)
          (ash x -1)
          (ash (logxor x gp) -1)))))

(define (make-lfsr-stream gp init stepper)
  (lfsr-iterator (stepper gp) init))

(define (make-lfsr-stream-fibonacci gp init)
  (make-lfsr-stream gp init lfsr-step-fibonacci))

(define (make-lfsr-stream-galois gp init)
  (make-lfsr-stream gp init lfsr-step-galois))

(define (word->bit-lfsr strm)
  (stream-map (lambda (x) (logand x 1)) strm))


scheme/communication/gps-ca.scm
;; Copyright © 2017 Frank Terbeck <ft@bewatermyfriend.org>
;;
;; This library  is free  software; you  can redistribute  it and/or  modify it
;; under the terms of the GNU Lesser General Public License as published by the
;; Free  Software Foundation;  either version  3 of  the License,  or (at  your
;; option) any later version.
;;
;; This library is distributed in the hope  that it will be useful, but WITHOUT
;; ANY  WARRANTY;  without even  the  implied  warranty of  MERCHANTABILITY  or
;; FITNESS FOR A PARTICULAR PURPOSE. See  the GNU Lesser General Public License
;; for more details.
;;
;; You should  have received a  copy of the  GNU Lesser General  Public License
;; along with  this library;  if not,  write to  the Free  Software Foundation,
;; Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
;;
;; Commentary:
;;
;; This module implements streams that  produce pseudo random gold chip sequen-
;; ces as used in  C/A of the GPS satellite system. The  module uses the linear
;; feedback shift register module (communication lfsr)

(define-module (communication gps-ca)
  #:use-module (srfi srfi-41)
  #:use-module (communication lfsr)
  #:export (make-gps-ca-stream
            gps-prn-id->code-delay))

(define stream-g1-parallel (make-lfsr-stream-fibonacci #b10010000001 #b1111111111))
(define stream-g1-serial (word->bit-lfsr stream-g1-parallel))
(define stream-g2-parallel (make-lfsr-stream-fibonacci #b10110010111 #b1111111111))

(define (gen-selector a b)
  (logior (ash 1 (- 9 a))
          (ash 1 (- 9 b))))

(define gps-prn-id-config
  (vector (cons (gen-selector 1 5) 5)
          (cons (gen-selector 2 6) 6)
          (cons (gen-selector 3 7) 7)
          (cons (gen-selector 4 8) 8)
          (cons (gen-selector 0 8) 17)
          (cons (gen-selector 1 9) 18)
          (cons (gen-selector 0 7) 139)
          (cons (gen-selector 1 8) 140)
          (cons (gen-selector 2 9) 141)
          (cons (gen-selector 1 2) 251)
          (cons (gen-selector 2 3) 252)
          (cons (gen-selector 4 5) 254)
          (cons (gen-selector 5 6) 255)
          (cons (gen-selector 6 7) 256)
          (cons (gen-selector 7 8) 257)
          (cons (gen-selector 8 9) 258)
          (cons (gen-selector 0 3) 469)
          (cons (gen-selector 1 4) 470)
          (cons (gen-selector 2 5) 471)
          (cons (gen-selector 3 6) 472)
          (cons (gen-selector 4 7) 473)
          (cons (gen-selector 5 8) 474)
          (cons (gen-selector 0 2) 509)
          (cons (gen-selector 3 5) 512)
          (cons (gen-selector 4 6) 513)
          (cons (gen-selector 5 7) 514)
          (cons (gen-selector 6 8) 515)
          (cons (gen-selector 7 9) 516)
          (cons (gen-selector 0 5) 859)
          (cons (gen-selector 1 6) 860)
          (cons (gen-selector 2 7) 861)
          (cons (gen-selector 3 8) 862)
          (cons (gen-selector 4 9) 863)
          (cons (gen-selector 3 9) 950)
          (cons (gen-selector 0 6) 947)
          (cons (gen-selector 1 7) 948)
          (cons (gen-selector 3 9) 950)))

(define (gps-prn-id->phase-selector n)
  (car (vector-ref gps-prn-id-config (- n 1))))

(define (gps-prn-id->code-delay n)
  (cdr (vector-ref gps-prn-id-config (- n 1))))

(define (make-g2-serial-stream gps-prn-id)
  (let ((phase-select (gps-prn-id->phase-selector gps-prn-id)))
    (stream-map (lambda (x) (multi-xor (logand x phase-select)))
                stream-g2-parallel)))

(define (make-gps-ca-stream gps-prn-id)
  (stream-map (lambda (g1 g2) (logxor g1 g2))
              stream-g1-serial
              (make-g2-serial-stream gps-prn-id)))


scheme/communication/ieee802-15-4a-preamble.scm
;; Copyright © 2017 Frank Terbeck <ft@bewatermyfriend.org>
;;
;; This library  is free  software; you  can redistribute  it and/or  modify it
;; under the terms of the GNU Lesser General Public License as published by the
;; Free  Software Foundation;  either version  3 of  the License,  or (at  your
;; option) any later version.
;;
;; This library is distributed in the hope  that it will be useful, but WITHOUT
;; ANY  WARRANTY;  without even  the  implied  warranty of  MERCHANTABILITY  or
;; FITNESS FOR A PARTICULAR PURPOSE. See  the GNU Lesser General Public License
;; for more details.
;;
;; You should  have received a  copy of the  GNU Lesser General  Public License
;; along with  this library;  if not,  write to  the Free  Software Foundation,
;; Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
;;
;; Commentary:
;;
;; This module  implements ternary  sequences as  used in  ieee-802.15.4a ultra
;; wideband  beacon  preambles.  The  module uses  the  linear  feedback  shift
;; register module (communication lfsr) with LFSRs in Galois configuration.

(define-module (communication ieee802-15-4a-preamble)
  #:use-module (srfi srfi-41)
  #:use-module (communication lfsr)
  #:export (ternary-code-chips
            ternary-code-31-chips
            ternary-code-127-chips))

(define (ternary-mapper n)
  (if (zero? n) 1 -1))

(define ternary-code-31-cfg
  (vector (vector 41 20    61 20)
          (vector 47  8    59  1)
          (vector 41 29    59 29)
          (vector 61  1    41 27)
          (vector 37 20    61 19)
          (vector 61 16    59  3)
          (vector 59 29    61  6)
          (vector 55 12    41 18)))

(define (fetch-31 k n)
  (vector-ref (vector-ref ternary-code-31-cfg (- n 1)) k))

(define (lfsr-a-31-gp n)   (fetch-31 0 n))
(define (lfsr-a-31-init n) (fetch-31 1 n))
(define (lfsr-b-31-gp n)   (fetch-31 2 n))
(define (lfsr-b-31-init n) (fetch-31 3 n))

(define (make-code-31-a-stream n)
  (make-lfsr-stream-galois (lfsr-a-31-gp n) (lfsr-a-31-init n)))

(define (make-code-31-b-stream n)
  (make-lfsr-stream-galois (lfsr-b-31-gp n) (lfsr-b-31-init n)))

(define (ternary-code-31-chip-stream n)
  (let* ((stream-a (word->bit-lfsr (make-code-31-a-stream n)))
         (stream-b (word->bit-lfsr (make-code-31-b-stream n))))
    (stream-map (lambda (a b)
                  (* a (ternary-mapper b)))
                stream-a
                stream-b)))

(define (ternary-code-31-chips n)
  (reverse (stream->list 31 (ternary-code-31-chip-stream n))))


(define ternary-code-127-cfg
  (vector (vector 131  73    143  76    185 122)
          (vector 137  99    143 106    157 101)
          (vector 241  90    247 123    253  31)
          (vector 167 117    143  42    137 109)
          (vector 157  72    203 100    239  70)
          (vector 203  96    143   1    191  50)
          (vector 247  32    131 106    145 124)
          (vector 171 126    185 115    191 122)
          (vector 193  73    157  47    241  25)
          (vector 145  99    185  83    241  43)
          (vector 143  90    191  71    239  49)
          (vector 229  87    241  42    145  91)
          (vector 185  72    211  75    247 102)
          (vector 211  48    241  76    253 110)
          (vector 239   8    137 124    193 109)
          (vector 213 126    157  83    253  94)))

(define (fetch-127 k n)
  (vector-ref (vector-ref ternary-code-127-cfg (- n 1)) k))

(define (lfsr-a-127-gp n)   (fetch-127 0 n))
(define (lfsr-a-127-init n) (fetch-127 1 n))
(define (lfsr-b-127-gp n)   (fetch-127 2 n))
(define (lfsr-b-127-init n) (fetch-127 3 n))
(define (lfsr-c-127-gp n)   (fetch-127 4 n))
(define (lfsr-c-127-init n) (fetch-127 5 n))

(define (make-code-127-a-stream n)
  (make-lfsr-stream-galois (lfsr-a-127-gp n) (lfsr-a-127-init n)))

(define (make-code-127-b-stream n)
  (make-lfsr-stream-galois (lfsr-b-127-gp n) (lfsr-b-127-init n)))

(define (make-code-127-c-stream n)
  (make-lfsr-stream-galois (lfsr-c-127-gp n) (lfsr-c-127-init n)))

(define (ternary-code-127-chip-stream n)
  (let* ((stream-a (word->bit-lfsr (make-code-127-a-stream n)))
         (stream-b (word->bit-lfsr (make-code-127-b-stream n)))
         (stream-c (word->bit-lfsr (make-code-127-c-stream n))))
    (stream-map (lambda (a b c)
                  (* a (ternary-mapper (logxor b c))))
                stream-a
                stream-b
                stream-c)))

(define (ternary-code-127-chips n)
  (reverse (stream->list 127 (ternary-code-127-chip-stream n))))

(define (ternary-code-chips n)
  (let* ((short? (< n 9))
         (size (or (and short? 31) 127))
         (stream (or (and short? ternary-code-31-chip-stream)
                     ternary-code-127-chip-stream))
         (index (or (and short? n) (- n 8))))
    (reverse (stream->list size (stream index)))))
